### ISSUE TITLE
Mask usage

### ISSUE THREAD

**Issue by kkirtac** [yabufarha/ms-tcn#22]
Hi @yabufarha ,

I understand that you are maintaining a binary mask to avoid performing computations for the padded entries.

The second dimension of your mask is indexing class labels:
`mask = torch.zeros(len(batch_input), self.num_classes, max(length_of_sequences), dtype=torch.float)`

I noticed you are indexing only the first class label in your model's forward pass,
[DilatedResidualLayer](https://github.com/yabufarha/ms-tcn/blob/master/model.py#L52), 
[SingleStageModel](https://github.com/yabufarha/ms-tcn/blob/master/model.py#L37),
[MultiStageModel ](https://github.com/yabufarha/ms-tcn/blob/master/model.py#L21),

Could you please explain the reason?

Many thanks.

**Comment by yabufarha** [comment]
No actually here I'm using the broadcasting property in pytorch. It will mask all the elements of that dimension. Maybe I should have set the second dimension of the mask to 1 to avoid confusion.

Just to confirm, I'm masking all elements in the second dimension not only the first one. I'm using [:, 0:1, :] to index the mask to make sure that the 2nd dimension is 1, which is required for broadcasting. 
I'm relying on broadcasting as all elements in that dimension have the same mask, and this mask is used to mask predicted probabilities (see the smoothing loss) where the second dimension equals num_class, and also to mask features (as in the lines you referred to) where the second dimension equals the number of feature maps.

I hope this would resolve the confusion.

**Comment by kkirtac** [comment]
Not sure if I understand correctly. To clarify it with an example,
Suppose I have two arrays with different lengths, and I want to pad them to same length. And suppose I have a total 2 class labels in my dataset.

`label1 = torch.LongTensor(list(range(10)))`
`label2 = torch.LongTensor(list(range(5)))`
`labels = [label1, label2]`
`labels_padded = torch.nn.utils.rnn.pad_sequence(labels, batch_first=True, padding_value=-1)`
`print(labels_padded.shape)`
`num_classes = 2`
`mask = torch.zeros(len(labels_padded), num_classes , labels_padded.shape[1], dtype=torch.float)`
```
for i in range(num_classes):
    mask[:, i, :] = (labels_padded != -1)
```
`print(mask.shape)`

Would this produce a mask identical to yours and would serve for the same purpose?
